File: src/spline.c
#include "assets.h"
#include "collision.h"
#include "linalg.h"
#include "macros.h"
#include "render.h"
#include "spline.h"

#include <math.h>

#define BAKE_LENGTH_SQ 1.0f

#define FORWARD_VEC_SIZE 0.125f

static void bezier(const Spline *spline, uint8_t index, float offset, Vec v) {
    uint8_t otherIndex = (index + 2) % spline->numPoints;
    vec_scaled_copy(v, spline->points[index].point, (1.0f - offset) * (1.0f - offset));
    vec_scaled_add(v, spline->points[index].control, 2.0f * (1.0f - offset) * offset);
    vec_scaled_add(v, spline->points[otherIndex].point, offset * offset);
}

static void interpolate(const Spline *spline, float offset, Vec v) {
    uint8_t index = offset;
    offset -= index;
    index %= spline->numPoints;
    uint8_t prevIndex = (index + spline->numPoints - 1) % spline->numPoints;
    float prevMid = spline->points[prevIndex].controlMid;
    float nextMid = spline->points[index].controlMid;
    Vec temp;
    bezier(spline, prevIndex, offset * (1.0f - prevMid) + prevMid, v);
    bezier(spline, index, offset * nextMid, temp);
    vec_scale(v, 1.0f - offset);
    vec_scaled_add(v, temp, offset);
}

static void add_baked(Spline *spline, float position) {
    SplineBaked *baked = &spline->baked[spline->numBaked];
    baked->position = position;
    interpolate(spline, position, baked->point);
    if (spline->numBaked != 0) {
        spline->length += sqrtf(vec_distance_sq(baked->point, spline->baked[spline->numBaked - 1].point));
    }
    baked->offset = spline->length;
    ++spline->numBaked;
}

static void bake_recursive(Spline *spline, int index, float begin, float end, int depth) {
    if (depth >= MAX_BAKE_DEPTH) return;

    Vec v1, v2;
    interpolate(spline, index + begin, v1);
    interpolate(spline, index + end, v2);

    float segment_length_squared = vec_distance_sq(v1, v2);
    if (segment_length_squared > BAKE_LENGTH_SQ) {
        float mid = (begin + end) * 0.5f;
        // recurse on either end of midpoint - in-order to avoid sorting
        bake_recursive(spline, index, begin, mid, depth + 1);
        add_baked(spline, index + mid);
        bake_recursive(spline, index, mid, end, depth + 1);
    }
}

static void bake(Spline *spline) {
    // set starting size of spline to zero
    spline->length = 0.0f;
    // index to write baked points to
    spline->numBaked = 0;
    // for each point, recursively find points to bake
    for (uint8_t i = 0; i < spline->numPoints; i++) {
        // bake at control point
        add_baked(spline, i);
        // add length to tilt offsets
        spline->points[i].tiltOffset = spline->length;
        // bake in between
        bake_recursive(spline, i, 0.0f, 1.0f, 0);
    }
    // finish off length measurement
    spline->length += sqrtf(vec_distance_sq(spline->baked[0].point, spline->baked[spline->numBaked - 1].point));
}

static void generate_controls(Spline *spline) {
    // generate bezier control points
    for (uint8_t a = 0; a < spline->numPoints; a++) {
        uint8_t b = (a + 1) % spline->numPoints;
        uint8_t c = (a + 2) % spline->numPoints;
        float da = sqrtf(vec_distance_sq(spline->points[a].point, spline->points[b].point));
        float db = sqrtf(vec_distance_sq(spline->points[b].point, spline->points[c].point));
        // TODO handle potential divs by zero in this area
        float mid = da / (da + db);
        float fac_a = (mid - 1.0f) / (2.0f * mid);
        float fac_b = 1.0f / (2.0f * mid * (1.0f - mid));
        float fac_c = mid / (2.0f * (mid - 1.0f));
        vec_scaled_copy(spline->points[a].control, spline->points[a].point, fac_a);
        vec_scaled_add(spline->points[a].control, spline->points[b].point, fac_b);
        vec_scaled_add(spline->points[a].control, spline->points[c].point, fac_c);
        spline->points[a].controlMid = mid;
    }
}

bool spline_load(Spline *spline, Asset *asset) {
    // static buffers to hold data before building spline
    // this allows verifying input without modifying the existing spline
    // so if this function fails, no invaraints are invalidated
    static uint8_t numPoints;
    static Vec points[MAX_POINTS];
    static float tilts[MAX_POINTS];
    // number of points
    if (!asset_read_byte(asset, &numPoints)) return false;
    if (numPoints < 3 || numPoints > MAX_POINTS) return false;
    // read points in
    for (int i = 0; i < numPoints; i++) {
        if (!asset_read_vec(asset, points[i])) return false;
        uint8_t tilt_int;
        if (!asset_read_byte(asset, &tilt_int)) return false;
        tilts[i] = (tilt_int / 256.0f) * (2.0f * PI);
    }
    // TODO handle div by zero tests
    // otherwise data looks good
    spline->numPoints = numPoints;
    for (uint8_t i = 0; i < numPoints; i++) {
        vec_copy(spline->points[i].point, points[i]);
        spline->points[i].tilt = tilts[i];
    }
    // fix tilts
    spline->totalTilt = spline->points[0].tilt;
    for (int i = 0; i < spline->numPoints; i++) {
        float delta = fmodf(spline->points[(i + 1) % spline->numPoints].tilt - spline->points[i].tilt + 2.0f * PI, 2.0f * PI);
        spline->points[i].tilt = spline->totalTilt;
        if (delta <= PI) {
            // move up
            spline->totalTilt += delta;
        } else {
            // move down
            spline->totalTilt += delta - (2.0f * PI);
        }
    }
    generate_controls(spline);
    bake(spline);
    return true;
}

static float convert_baked_offset(const Spline *spline, float baked_offset) {
    // binary search
    size_t start = 0;
    size_t end = spline->numBaked;
    size_t current = (start + end) / 2;
    while (start < current) {
        if (baked_offset <= spline->baked[current].offset) {
            end = current;
        } else {
            start = current;
        }
        current = (start + end) / 2;
    }
    // interpolate
    float offsetBegin = spline->baked[current].offset;
    float offsetEnd = spline->baked[(current + 1) % spline->numBaked].offset;
    float positionBegin = spline->baked[current].position;
    float positionEnd = spline->baked[(current + 1) % spline->numBaked].position;
    if (current == spline->numBaked - 1) {
        offsetEnd += spline->length;
        positionEnd += spline->numPoints;
    }
    float interp = (baked_offset - offsetBegin) / (offsetEnd - offsetBegin);
    float result = (1.0f - interp) * positionBegin + interp * positionEnd;
    return result;
}

void spline_get_baked(const Spline *spline, float offset, Vec v) {
    interpolate(spline, convert_baked_offset(spline, offset), v);
}

static float get_tilt_offset(const Spline *spline, int i) {
    int n = i / spline->numPoints;
    return spline->length * n + spline->points[i - n * spline->numPoints].tiltOffset;
}

static float get_tilt_radian(const Spline *spline, int i) {
    int n = i / spline->numPoints;
    return spline->totalTilt * n + spline->points[i - n * spline->numPoints].tilt;
}

static float lagrange(const Spline *spline, int i, float x) {
    // TODO optimize
    float x0 = get_tilt_offset(spline, i);
    float x1 = get_tilt_offset(spline, i + 1);
    float x2 = get_tilt_offset(spline, i + 2);
    float y0 = get_tilt_radian(spline, i);
    float y1 = get_tilt_radian(spline, i + 1);
    float y2 = get_tilt_radian(spline, i + 2);
	float result = (y0 * (x - x1) / (x0 - x1) * (x - x2) / (x0 - x2));
	result += (y1 * (x - x0) / (x1 - x0) * (x - x2) / (x1 - x2));
	result += (y2 * (x - x0) / (x2 - x0) * (x - x1) / (x2 - x1));
    return result;
}

float spline_get_tilt(const Spline *spline, float offset) {
    float pre_baked = fmodf(offset, spline->length);
    offset = convert_baked_offset(spline, offset);
    int index = offset;
    float a = lagrange(spline, index - 1, pre_baked);
    float b = lagrange(spline, index, pre_baked);
    offset -= index;
    return a * (1.0f - offset) + b * offset;
}

void spline_get_up_right(const Spline *spline, float offset, Vec up, Vec right) {
    float sa = fmodf(offset - FORWARD_VEC_SIZE + spline->length, spline->length);
    float sb = fmodf(offset + FORWARD_VEC_SIZE + spline->length, spline->length);
    Vec target, temp;
    spline_get_baked(spline, sb, target);
    spline_get_baked(spline, sa, temp);
    vec_sub(target, temp);
    vec_normalize(target);
    Mtx look, rot;
    // TODO would have issues for track going directly upwards
    mtx_look_at(look, target, gVecYAxis);
    float tilt = spline_get_tilt(spline, offset);
    if (up != NULL) {
        mtx_mul_vec(look, temp, gVecYAxis);
        mtx_angle_axis(rot, target, tilt);
        mtx_mul_vec(rot, up, temp);
    }
    if (right != NULL) {
        mtx_mul_vec(look, temp, gVecXAxis);
        mtx_angle_axis(rot, target, tilt);
        mtx_mul_vec(rot, right, temp);
    }
}

static void get_distance(const Spline *spline, const Vec point, int i, float *distance, float *nearest) {
    float offset = spline->baked[i].offset;
    float interval = spline->baked[(i + 1) % spline->numBaked].offset - offset;
    Vec origin;
    vec_copy(origin, spline->baked[i].point);
    Vec direction;
    vec_copy(direction, spline->baked[(i + 1) % spline->numBaked].point);
    vec_sub(direction, origin);
    vec_scale(direction, 1.0f / interval);
    Vec proj;
    vec_copy(proj, point);
    vec_sub(proj, origin);
    float d = vec_dot(proj, direction);
    if (d < 0.0f) d = 0.0f;
    else if (d > interval) d = interval;
    vec_copy(proj, direction);
    vec_scale(proj, d);
    vec_add(proj, origin);
    float dist = vec_distance_sq(proj, point);
    if (dist < *distance) {
        *nearest = offset + d;
        *distance = dist;
    }
}

static float get_closest(const Spline *spline, const Octree *tree, const Vec point) {
    Vec min, max;
    vec_copy(min, tree->min);
    vec_copy(max, tree->max);

    const OctreeNode *current = &tree->root;
    float distance = INFINITY;
    float nearest = 0.0f;

    for (;;) {
        int which[3];
        octree_find_which(point, min, max, which);

        int segment = current->segments;
        while (segment >= 0) {
            if (!(which[0] == 1 && (tree->segmentSides[segment] & 1)) &&
                !(which[0] == 0 && (tree->segmentSides[segment] & 2)) &&
                !(which[1] == 1 && (tree->segmentSides[segment] & 4)) &&
                !(which[1] == 0 && (tree->segmentSides[segment] & 8)) &&
                !(which[2] == 1 && (tree->segmentSides[segment] & 16)) &&
                !(which[2] == 0 && (tree->segmentSides[segment] & 32))) {
                get_distance(spline, point, segment, &distance, &nearest);
            }
            segment = tree->segmentNext[segment];
        }

        if (current->children_index == -1) {
            break;
        }
        current = &tree->childPool[current->children_index + (which[0] | (which[1] << 1) | (which[2] << 2))];
    }

    return nearest;
}

bool spline_get_up_height(const Spline *spline, const Octree *tree, const Vec pos, Vec up, float *height) {
    float offset = get_closest(spline, tree, pos);
    Vec point;
    spline_get_baked(spline, offset, point);
    Vec right;
    spline_get_up_right(spline, offset, up, right);
    Vec d;
    vec_copy(d, pos);
    vec_sub(d, point);
    float side_distance = vec_dot(right, d);
    if (side_distance < -SPLINE_TRACK_RADIUS || side_distance > SPLINE_TRACK_RADIUS) {
        return false;
    }
    *height = vec_dot(up, d);
    return true;
}

File: src/spline.rs
//! Condux - an antigravity racing game
//! Copyright (C) 2023 spazzylemons
//!
//! This program is free software: you can redistribute it and/or modify
//! it under the terms of the GNU General Public License as published by
//! the Free Software Foundation, either version 3 of the License, or
//! (at your option) any later version.
//!
//! This program is distributed in the hope that it will be useful,
//! but WITHOUT ANY WARRANTY; without even the implied warranty of
//! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//! GNU General Public License for more details.
//!
//! You should have received a copy of the GNU General Public License
//! along with this program.  If not, see <http://www.gnu.org/licenses/>.

use std::{
    f32::consts::{PI, TAU},
    sync::Arc,
};

use crate::{
    assets::Asset,
    linalg::{Length, Mtx, Vector},
    octree::Octree,
    render::graph::RenderGraph3d,
    vehicle::Vehicle,
};

const BAKE_LENGTH_SQ: f32 = 1.0;

const FORWARD_VECTOR_SIZE: f32 = 0.125;

struct Point {
    point: Vector,
    control: Vector,
    control_mid: f32,
    tilt: f32,
    tilt_offset: f32,
}

pub struct Baked {
    pub point: Vector,
    position: f32,
    pub offset: f32,
}

pub struct Spline {
    /// The control points.
    points: Vec<Point>,
    /// The baked points.
    pub baked: Vec<Baked>,
    /// The total tilt, used for interpolation.
    total_tilt: f32,
    /// The approximate length of the spline.
    pub length: f32,

    /// The points to render for the floor.
    render_floor: Arc<Vec<(Vector, Vector)>>,
    /// The points to render for the walls.
    render_walls: Arc<Vec<(Vector, Vector)>>,
}

pub struct CollisionState {
    /// The up vector.
    pub up: Vector,
    /// The right vector.
    pub right: Vector,
    /// The height above (or slightly) below the spline.
    pub height: f32,
    /// The horizontal offset on the spline.
    pub horizontal: f32,
    /// The offset of the closest point.
    pub offset: f32,
}

impl Spline {
    /// The radius of the spline.
    pub const TRACK_RADIUS: f32 = 2.0;
    /// The extended spline radius for bounds checks.
    pub const BOUNDS_RADIUS: f32 = 5.0;
    /// The height of the walls on the side.
    pub const WALL_HEIGHT: f32 = 0.25;

    const MAX_BAKE_DEPTH: usize = 5;

    pub fn load(asset: &mut Asset) -> Option<Self> {
        // number of points
        let num_points = asset.read_byte()?;
        if num_points < 3 {
            return None;
        }
        let mut points = vec![];
        // TODO div by zero checks?
        for _ in 0..num_points {
            let point = asset.read_vector()?;
            let tilt = (f32::from(asset.read_byte()?) / 256.0) * TAU;
            points.push(Point {
                point,
                control: Vector::default(),
                control_mid: 0.0,
                tilt,
                tilt_offset: 0.0,
            });
        }
        // fix tilts
        let mut total_tilt = points[0].tilt;
        for i in 0..num_points {
            let delta = (points[usize::from((i + 1) % num_points)].tilt
                - points[usize::from(i)].tilt)
                .rem_euclid(TAU);
            points[usize::from(i)].tilt = total_tilt;
            if delta <= PI {
                // move up
                total_tilt += delta;
            } else {
                // move down
                total_tilt += delta - TAU;
            }
        }
        // generate bezier control points
        for i in 0..num_points {
            let a = usize::from(i);
            let b = usize::from((i + 1) % num_points);
            let c = usize::from((i + 2) % num_points);
            let pa = points[a].point;
            let pb = points[b].point;
            let pc = points[c].point;
            let da = pa.dist(pb);
            let db = pb.dist(pc);
            // TODO handle potential divs by zero in this area
            let mid = da / (da + db);
            let fac_a = (mid - 1.0) / (2.0 * mid);
            let fac_b = 1.0 / (2.0 * mid * (1.0 - mid));
            let fac_c = mid / (2.0 * (mid - 1.0));
            points[a].control = (pa * fac_a) + (pb * fac_b) + (pc * fac_c);
            points[a].control_mid = mid;
        }
        let mut spline = Self {
            points,
            baked: vec![],
            total_tilt,
            length: 0.0,

            render_floor: Arc::new(vec![]),
            render_walls: Arc::new(vec![]),
        };
        // for each point, recursively find points to bake
        for i in 0..num_points {
            // bake at control point
            spline.add_baked(f32::from(i));
            // add length to tilt offsets
            spline.points[usize::from(i)].tilt_offset = spline.length;
            // bake in between
            spline.bake_recursive(i, 0.0, 1.0, 0);
        }
        // finish off length measurement
        let final_length = spline.baked[0]
            .point
            .dist(spline.baked[spline.baked.len() - 1].point);
        spline.length += final_length;
        // build render info
        spline.prerender();
        // all good
        Some(spline)
    }

    #[must_use]
    #[allow(clippy::cast_possible_truncation)]
    fn num_points(&self) -> u8 {
        // the number of points is always known to be within u8 range
        self.points.len() as u8
    }

    #[must_use]
    pub fn bezier(&self, index: usize, offset: f32) -> Vector {
        let other_index = (index + 2) % self.points.len();
        let fac_a = (1.0 - offset) * (1.0 - offset);
        let fac_b = 2.0 * (1.0 - offset) * offset;
        let fac_c = offset * offset;
        let pa = self.points[index].point;
        let pb = self.points[index].control;
        let pc = self.points[other_index].point;
        pa * fac_a + pb * fac_b + pc * fac_c
    }

    #[must_use]
    pub fn interpolate(&self, offset: f32) -> Vector {
        let offset = offset.rem_euclid(f32::from(self.num_points()));
        let index = offset as usize;
        let offset = offset - offset.floor();
        let prev_index = (index + self.points.len() - 1) % self.points.len();
        let prev_mid = self.points[prev_index].control_mid;
        let next_mid = self.points[index].control_mid;
        let a = self.bezier(prev_index, offset * (1.0 - prev_mid) + prev_mid);
        let b = self.bezier(index, offset * next_mid);
        a * (1.0 - offset) + b * offset
    }

    pub fn add_baked(&mut self, position: f32) {
        let baked = Baked {
            point: self.interpolate(position),
            position,
            offset: 0.0,
        };
        if !self.baked.is_empty() {
            self.length += baked.point.dist(self.baked[self.baked.len() - 1].point);
        }
        self.baked.push(Baked {
            offset: self.length,
            ..baked
        });
    }

    pub fn bake_recursive(&mut self, index: u8, begin: f32, end: f32, depth: usize) {
        if depth >= Self::MAX_BAKE_DEPTH {
            return;
        }

        let v1 = self.interpolate(f32::from(index) + begin);
        let v2 = self.interpolate(f32::from(index) + end);

        if v1.dist_sq(v2) > BAKE_LENGTH_SQ {
            let mid = (begin + end) * 0.5;
            self.bake_recursive(index, begin, mid, depth + 1);
            self.add_baked(f32::from(index) + mid);
            self.bake_recursive(index, mid, end, depth + 1);
        }
    }

    fn convert_baked_offset(&self, baked_offset: f32) -> f32 {
        // binary search
        let mut start = 0;
        let mut end = self.baked.len();
        let mut current = (start + end) / 2;
        while start < current {
            if baked_offset <= self.baked[current].offset {
                end = current;
            } else {
                start = current;
            }
            current = (start + end) / 2;
        }
        // interpolate
        let next_index = (current + 1) % self.baked.len();
        let offset_begin = self.baked[current].offset;
        let mut offset_end = self.baked[next_index].offset;
        let position_begin = self.baked[current].position;
        let mut position_end = self.baked[next_index].position;
        if next_index == 0 {
            offset_end += self.length;
            position_end += f32::from(self.num_points());
        }
        let interp = (baked_offset - offset_begin) / (offset_end - offset_begin);
        (1.0 - interp) * position_begin + interp * position_end
    }

    #[must_use]
    pub fn get_baked(&self, offset: f32) -> Vector {
        self.interpolate(self.convert_baked_offset(offset))
    }

    #[must_use]
    fn floor_div(&self, i: isize) -> (isize, &Point) {
        let n = isize::from(self.num_points());
        let d = i / n;
        let d = if i < 0 && d * i != n { d - 1 } else { d };
        (d, &self.points[(i - d * n) as usize])
    }

    fn get_tilt_offset(&self, i: isize) -> f32 {
        let (n, p) = self.floor_div(i);
        self.length * n as f32 + p.tilt_offset
    }

    fn get_tilt_radian(&self, i: isize) -> f32 {
        let (n, p) = self.floor_div(i);
        self.total_tilt * n as f32 + p.tilt
    }

    fn lagrange(&self, i: isize, x: f32) -> f32 {
        // TODO optimize
        let x0 = self.get_tilt_offset(i);
        let x1 = self.get_tilt_offset(i + 1);
        let x2 = self.get_tilt_offset(i + 2);
        let y0 = self.get_tilt_radian(i);
        let y1 = self.get_tilt_radian(i + 1);
        let y2 = self.get_tilt_radian(i + 2);
        (y0 * (x - x1) / (x0 - x1) * (x - x2) / (x0 - x2))
            + (y1 * (x - x0) / (x1 - x0) * (x - x2) / (x1 - x2))
            + (y2 * (x - x0) / (x2 - x0) * (x - x1) / (x2 - x1))
    }

    fn get_tilt(&self, offset: f32) -> f32 {
        let pre_baked = offset.rem_euclid(self.length);
        let offset = self.convert_baked_offset(offset);
        let index = offset as isize;
        let a = self.lagrange(index - 1, pre_baked);
        let b = self.lagrange(index, pre_baked);
        let offset = offset - offset.floor();
        a * (1.0 - offset) + b * offset
    }

    #[must_use]
    pub fn get_up_right(&self, offset: f32) -> (Vector, Vector) {
        let sa = (offset - FORWARD_VECTOR_SIZE).rem_euclid(self.length);
        let sb = (offset + FORWARD_VECTOR_SIZE).rem_euclid(self.length);
        let target = (self.get_baked(sb) - self.get_baked(sa)).normalized();
        let look = Mtx::looking_at(target, Vector::Y_AXIS);
        let tilt = Mtx::axis_angle(&target, self.get_tilt(offset));
        let up = (Vector::Y_AXIS * look) * tilt;
        let right = (Vector::X_AXIS * look) * tilt;
        (up, right)
    }

    #[must_use]
    pub fn get_collision(&self, octree: &Octree, pos: Vector) -> Option<CollisionState> {
        let offset = if let Some(offset) = octree.find_closest_offset(self, pos) {
            offset
        } else {
            return None;
        };
        let point = self.get_baked(offset);
        let (up, right) = self.get_up_right(offset);
        let d = pos - point;
        let horizontal = right.dot(&d);
        let radius = horizontal.abs();
        if radius > Self::BOUNDS_RADIUS {
            // bounds radius check
            None
        } else {
            let height = up.dot(&d);
            // collision height check
            if (-Vehicle::COLLISION_DEPTH..=Vehicle::MAX_GRAVITY_HEIGHT).contains(&height) {
                Some(CollisionState {
                    up,
                    right,
                    height,
                    horizontal,
                    offset,
                })
            } else {
                None
            }
        }
    }

    #[must_use]
    pub fn get_offset_and_dist_sq(&self, point: Vector, index: usize) -> (f32, f32) {
        let next_index = (index + 1) % self.baked.len();
        let offset = self.baked[index].offset;
        let interval = (self.baked[next_index].offset - offset).abs();
        let origin = self.baked[index].point;
        let direction = (self.baked[next_index].point - origin) / interval;
        let proj = point - origin;
        let d = proj.dot(&direction).clamp(0.0, interval);
        let proj = (direction * d) + origin;
        let dist = proj.dist_sq(point);
        (offset + d, dist)
    }

    fn prerender(&mut self) {
        let mut d = 0.0;
        let mut render_floor = vec![];
        let mut render_walls = vec![];
        while d < self.length {
            let p = self.get_baked(d);
            let (mut u, r) = self.get_up_right(d);
            let r = r * Self::TRACK_RADIUS;
            let mut pl = p - r;
            let mut pr = p + r;
            render_floor.push((pl, pr));
            u *= Self::WALL_HEIGHT;
            pl += u;
            pr += u;
            render_walls.push((pl, pr));
            d += 1.0;
        }
        // in case no points loaded, don't draw
        if render_floor.is_empty() {
            return;
        }

        let mut my_render_floor = vec![];
        let mut my_render_walls = vec![];
        for index in 0..render_floor.len() {
            // avoid modulus for efficiency
            let mut other_index = index + 1;
            if other_index == render_floor.len() {
                other_index = 0;
            }
            // get floor points
            let (l1, r1) = render_floor[index];
            let (l2, r2) = render_floor[other_index];
            my_render_floor.push((l1, l2));
            my_render_floor.push((r1, r2));
            my_render_floor.push((l1, r1));
            // get wall points
            let (wl1, wr1) = render_walls[index];
            let (wl2, wr2) = render_walls[other_index];
            my_render_walls.push((l1, wl1));
            my_render_walls.push((r1, wr1));
            my_render_walls.push((wl1, wl2));
            my_render_walls.push((wr1, wr2));
        }
        self.render_floor = Arc::new(my_render_floor);
        self.render_walls = Arc::new(my_render_walls);
    }

    pub fn render(&self, graph: &mut RenderGraph3d, walls: bool) {
        graph.lines(self.render_floor.clone());
        if walls {
            graph.lines(self.render_walls.clone());
        }
    }
}
